drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,230 @@
+#define AXP192_REG_INTER_TEMP      0x5E
+#define AXP192_REG_BAT_VOL         0x78
+#define AXP192_REG_BAT_CHG_CUR     0x7A
+#define AXP192_REG_BAT_DCHG_CUR    0x7C
+
+/* X-Powers AXP192 voltage sources */
+enum axp192_voltage_sources {
+	AXP192_SOURCE_BUCK1,
+	AXP192_SOURCE_BUCK2,
+	AXP192_SOURCE_BUCK3,
+	AXP192_SOURCE_LDO1,
+	AXP192_SOURCE_LDO2,
+	AXP192_SOURCE_LDO3
+};
+
+struct regulator_axp192_pconfig {
+	struct i2c_dt_spec i2c;
+};
+
+struct regulator_axp192_config {
+	struct regulator_common_config common;
+	const struct device *p;
+	uint8_t source;
+	int32_t init_uv;
+	struct gpio_dt_spec enable_gpios;
+};
+
+struct regulator_axp192_data {
+	struct regulator_common_data data;
+};
+
+static int buck_output_enable(const struct device *dev, uint8_t chan, uint8_t enable)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	uint8_t config_bit[3] = {0, 4, 1};
+	uint8_t reg_value = 0;
+
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, &reg_value);
+
+	if (enable) {
+		reg_value |= BIT(config_bit[chan]);
+	} else {
+		reg_value &= ~BIT(config_bit[chan]);
+	}
+	return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, reg_value);
+}
+
+static int buck_set_voltage(const struct device *dev, uint8_t chan, int32_t init_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	switch (chan) {
+	case AXP192_SOURCE_BUCK1:
+		return buck_output_enable(dev, chan, true);
+	case AXP192_SOURCE_BUCK2:
+		return buck_output_enable(dev, chan, true);
+	case AXP192_SOURCE_BUCK3:
+		return buck_output_enable(dev, chan, true);
+	default:
+		return -ENODEV;
+	}
+}
+
+static int ldo_output_enable(const struct device *dev, uint8_t chan, uint8_t enable)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	uint8_t config_bit[3] = {7, 2, 3};
+	uint8_t reg_value = 0;
+
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, &reg_value);
+
+	if (enable) {
+		reg_value |= BIT(config_bit[chan - 3]);
+	} else {
+		reg_value &= ~BIT(config_bit[chan - 3]);
+	}
+	return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, reg_value);
+}
+
+static int ldo_set_voltage(const struct device *dev, uint8_t chan, int32_t init_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	if (init_uv < 1800000 || init_uv > 3300000) {
+		return -EINVAL;
+	}
+
+	uint8_t reg_value = 0;
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, &reg_value);
+
+	switch (chan) {
+	case AXP192_SOURCE_LDO1:
+		return 0;
+	case AXP192_SOURCE_LDO2:
+		reg_value |= (((init_uv - 180000) / 100000) & 0x0F) << 4;
+		i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, reg_value);
```

Comment:
You may be able to use i2c_reg_update_byte_dt here.
You should also consider using the linear_range API for calculating the correct register value.